### PR TITLE
Serializing data with NumPy optional

### PR DESCRIPTION
**Pull request by mistakevin**
Resolves issue mentioned in PR https://github.com/bokeh/bokeh/pull/2594

### CODE REVIEW

File: bokeh/util/serialization.py [hunk 1]
```diff
@@ -190,16 +196,27 @@ def transform_numerical_array(obj):
...
-    """recursively dig until a flat array is found
-    convert it to a numpy array and send off to transform_array()
-    to handle nan, inf, -inf
+    """recursively dig until a flat list is found
+    if numpy is available convert the flat list to a numpy array
+    and send off to transform_array() to handle nan, inf, -inf
+    otherwise iterate through items in array converting non-json items
     """
-    if not any(isinstance(el, (list, tuple)) for el in datum):
+    if not any(isinstance(el, (list, tuple)) for el in datum) and is_numpy:
```
Comment (mattpap) [inline bokeh/util/serialization.py]: Conditions should be reversed. It's cheaper to start with the other one.

**Comment by mattpap** [comment]
Do fast and slow branches generate the same JSON?

**Comment by mistakevin** [comment]
Yes, equal json, despite order and id names, is expected.

**Comment by damianavila** [comment]
>Yes, equal JSON is expected.

@mattpap remind us about tests for this kind of things (thanks Mateusz), then... can you add a test for that? :wink:

@mistakevin, if you can not do it now, I can merge this one as is (LGTM) right now but the test should be submitted and merged before the release build (Thursday)

**Comment by mistakevin** [comment]
@damianavila @mattpap @bryevdv Test added testing `traverse_data()`'s ability to handle non-jsonifiable objects. Are their other things that we want to test? I'm not sure how deep to go.

File: bokeh/util/serialization.py [hunk 2]
```diff
@@ -189,17 +195,30 @@ def transform_numerical_array(obj):
...
-    convert it to a numpy array and send off to transform_array()
-    to handle nan, inf, -inf
+def traverse_data(datum, use_numpy=True):
+    """recursively dig until a flat list is found
+    if numpy is available convert the flat list to a numpy array
+    and send off to transform_array() to handle nan, inf, -inf
+    otherwise iterate through items in array converting non-json items
     """
-    if not any(isinstance(el, (list, tuple)) for el in datum):
+    global is_numpy
```
Comment (damianavila) [inline bokeh/util/serialization.py]: mmm... I don't like globals...

what about 

```python
def traverse_data(datum, is_numpy,  use_numpy=True):

...

if is_numpy and use_numpy and not any(isinstance(el, (list, tuple)) for el in datum):
```

Also make clear in the docstring that you are using `use_numpy` for testing purposes... in fact, I think there is a way to mock the import... but I don't have play enough with that to make it quickly...

**Comment by mistakevin** [comment]
@damianavila updated. Nice call about the global variables, they are ugly. Still catching up on python. ;)

**Pull request merged**